sherpa/astro/xspec/tests/test_xspec_regrid.py

```diff
@@ -0,0 +1,639 @@
+
+    We test out both additive and multiplicative models.
+    """
+
+    from sherpa.astro import xspec as xs
+
+    infile = make_data_path(name)
+    tbl = xs.read_xstable_model('mod', infile)
+
+    assert tbl.ndim == 1
+    assert tbl.integrate == iflag
+
+    egrid = np.arange(0.5, 1.5, 0.1)
+    eg1 = egrid[:-1]
+    eg2 = egrid[1:]
+
+    exp = tbl(eg1, eg2)
+    assert (exp > 0).all()
+
+    ebase = np.arange(0.45, 1.55, 0.05)
+    rtbl = tbl.regrid(ebase[:-1], ebase[1:])
+    y = rtbl(eg1, eg2)
+
+    assert y == pytest.approx(exp, rel=tol)
+
+
+@requires_xspec
+@pytest.mark.parametrize("mname", ["wabs", "powerlaw"])
+def test_regrid_convolution_requires_bins(mname, xsmodel):
+    """Does the regrid method require lo,hi bins?
+
+    This just checks the current behavior, which is about what is
+    allowed and what warnings are created, not the returned values.
+
+    """
+
+    ebase = [0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9]
+
+    conv = xsmodel("cflux")
+    mdl = xsmodel(mname, "base")
+    cmdl = conv(mdl)
+
+    rmdl = cmdl.regrid(ebase)
+    with pytest.warns(FutureWarning) as record:
+        rmdl([0.57, 0.62, 0.66, 0.72])
+
+    assert len(record) == 3
+    assert record[0].message.args[0] == 'calc() requires pars,lo,hi arguments, sent 2 arguments'
+    assert record[1].message.args[0] == 'calc() requires pars,rhs,lo,hi arguments, sent 3 arguments'
+    assert record[2].message.args[0] == 'calc() requires pars,lo,hi arguments, sent 2 arguments'
+
+
+@requires_xspec
+@pytest.mark.parametrize('model,tol',
+                         [('powerlaw', 1e-3),
+                          ('wabs', 0.72)])  # was tol=0.31 at one point
+def test_regrid_convolution_single(model, tol, xsmodel):
+    """regriding an additive model or multiplicative.
+    """
+
+    ebase = np.arange(0.4, 10, 0.1)
+    egrid = np.arange(0.5, 10, 0.2)
+    eg1 = egrid[:-1]
+    eg2 = egrid[1:]
+
+    conv = xsmodel("cflux")
+    conv.emin = 1
+    conv.emax = 9
+
+    mdl = xsmodel(model)
+    cmdl = conv(mdl)
+
+    rmdl = cmdl.regrid(ebase[:-1], ebase[1:])
+
+    exp = cmdl(eg1, eg2)
+    assert (exp > 0).all()
+
+    y = rmdl(eg1, eg2)
+    assert y == pytest.approx(exp, rel=tol)
+
+
+@requires_xspec
+@pytest.mark.parametrize('name1,name2',
+                         [('wabs', 'powerlaw'),
+                          ('powerlaw', 'wabs')])
+def test_regrid_convolution_combined(name1, name2, xsmodel):
+    """regriding convolved(mdl1 * mdl2)
+
+    The tolerance is very large
+    """
+
+    ebase = np.arange(0.4, 10, 0.1)
+    egrid = np.arange(0.5, 10, 0.2)
+    eg1 = egrid[:-1]
+    eg2 = egrid[1:]
+
+    conv = xsmodel("cflux")
+    conv.emin = 1
+    conv.emax = 9
+
+    mdl1 = xsmodel(name1)
+    mdl2 = xsmodel(name2)
+    cmdl = conv(mdl1 * mdl2)
+
+    rmdl = cmdl.regrid(ebase[:-1], ebase[1:])
+
+    exp = cmdl(eg1, eg2)
+    assert (exp > 0).all()
+
+    y = rmdl(eg1, eg2)
+    assert y == pytest.approx(exp, rel=0.35)
```

Comment:
It's so large that I wonder if something is wrong here... but as long as essentially everything is expected to fail anyway, I won't worry.

Reply:
I am treating this as a regression test at the moment, but it is one of the things I have been trying to get to the bottom of with my work on the regrid code, but have not been able to come to a satisfactory story just yet.